### PR TITLE
Move and update "memory model" docs

### PR DESCRIPTION
**Pull request by gonzalobg**
This PR moves the "memory model" related docs to a separate page, so that they can be easily linked from the Programming Guide. 

It also updates the memory model docs.

It provides a new guarantee that we did not provide before, system-scope atomics are also atomic if the operation:

> it is a load or store that affects a naturally-aligned object of sizes `1`, `2`, `4`, or `8` bytes on mapped memory

The scripts to build the site have stopped working, so this PR also updates those.

The site layout is constrained to a 800px body width, which makes tables render poorly. 
This PR bumps this to 1000px here.

Syntax highlighting within HTML tables seems to be buggy with Jekyll.

### CODE REVIEW

**Comment by gonzalobg** [comment]
@daniellustig please review.

File: docs/extended_api/memory_model.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
+---
+parent: Extended API
+nav_order: 0
+---
+
+# Memory model
+
+Standard C++ presents a view that the cost to synchronize threads is uniform and low.
+
+CUDA C++ is different: the cost to synchronize threads grows as threads are further appart.
```
Comment (daniellustig) [inline docs/extended_api/memory_model.md:10]: apart

File: docs/extended_api/memory_model.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+It is low across threads within a block, but high across arbitrary threads in the system running on multiple GPUs and CPUs.
+
+To account for non-uniform thread synchronization costs that are not always low, CUDA C++ extends the standard C++ memory model and concurrency facilities in the `cuda::` namespace with **thread scopes**, retaining the syntax and semantics of standard C++ by default.
+
+## Threads
+
+In CUDA C++, memory can only be modified by threads, so any operation that modifies memory is being executed by a thread.
+
+Examples of threads of execution in CUDA C++ programs: GPU or CPU threads, threads in DMA engines executing asynchronous operations like [`memcpy_async`] or [`cudaMemcpyAsync`], threads in NICs executing RDMA operations, etc.
```
Comment (daniellustig) [inline docs/extended_api/memory_model.md:19]: NACK this new Threads subsection.  We shouldn't talk about DMA engines, NICs, etc. here, and we shouldn't claim that nothing outside a thread can modify memory.

File: docs/extended_api/memory_model.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+[mapped memory]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#mapped-memory
+[unified memory]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#um-unified-memory-programming-hd
+[CUDA programming guide]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html
+[`concurrentManagedAccess`]: https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html#structcudaDeviceProp_116f9619ccc85e93bc456b8c69c80e78b
+[`hostNativeAtomicSupported`]: https://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html#structcudaDeviceProp_1ef82fd7d1d0413c7d6f33287e5b6306f
+
+## Data Races
+
+Modify [intro.races paragraph 21] of ISO/IEC IS 14882 (the C++ Standard) as follows:
+> The execution of a program contains a data race if it contains two potentially concurrent conflicting actions, at least one of which is not atomic ***at a scope that includes the thread that performed the other operation***, and neither happens before the other, except for the special case for signal handlers described below. Any such data race results inundefined behavior. [...]
```
Comment (daniellustig) [inline docs/extended_api/memory_model.md:82]: in undefined

File: docs/extended_api/memory_model.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+<tr>
+<td markdown="span" align="center"> 
+**Thread 0 Block 0** 
+</td><td markdown="span" align="center"> 
+**Thread 0 Block 1** 
+</td>
+</tr>
+<tr>
+<td markdown="span">
+`x = 42;`<br>
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply docs/extended_api/memory_model.md:165]: The documentation system that libcudacxx uses appears to have a bug that causes this.

Filled: https://github.com/NVIDIA/libcudacxx/issues/302

File: docs/extended_api/memory_model.md (hunk 2 above)
Comment (gonzalobg) [reply docs/extended_api/memory_model.md:19]: This section does not claim "that nothing outside a thread can modify memory." but rather that those operations modify memory via a C++ thread.

I can modify this sentence to:

```suggestion
Examples of threads of execution in CUDA C++ programs: GPU threads, CPU threads, thread asynchronously copying memory for [`memcpy_async`] or [`cudaMemcpyAsync`], etc.
```

File: docs/extended_api/memory_model.md (hunk 1 above)
Comment (gonzalobg) [reply docs/extended_api/memory_model.md:10]:
```suggestion
CUDA C++ is different: the cost to synchronize threads grows as threads are further apart.
```

File: docs/extended_api/memory_model.md (hunk 3 above)
Comment (gonzalobg) [reply docs/extended_api/memory_model.md:82]: Fix typo: 
```suggestion
> The execution of a program contains a data race if it contains two potentially concurrent conflicting actions, at least one of which is not atomic ***at a scope that includes the thread that performed the other operation***, and neither happens before the other, except for the special case for signal handlers described below. Any such data race results in undefined behavior. [...]
```

File: docs/extended_api/memory_model.md (hunk 2 above)
Comment (gonzalobg) [reply docs/extended_api/memory_model.md:19]: Leaving sentence out: 

```suggestion
```

File: docs/extended_api/memory_model.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+## Threads
+
+In CUDA C++, memory can only be modified by threads, so any operation that modifies memory is being executed by a thread.
+
+
+[`memcpy_async`]: asynchronous_operations/memcpy_async.md
+[`cudaMemcpyAsync`]: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79
+
+
```
Comment (gonzalobg) [inline docs/extended_api/memory_model.md:23]: Fix: 

```suggestion
```

**Comment by wmaxey** [comment]
@gonzalobg Is this good to get in?

**Comment by gonzalobg** [comment]
Waiting for review.

File: docs/extended_api/memory_model.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+## Synchronization primitives
+
+Types in namespaces `std::` and `cuda::std::` have the same behavior as corresponding types in namespace `cuda::` when instantiated with a scope of `cuda::thread_scope_system`.
+
+## Atomicity
+
+An atomic operation is atomic at the scope it specifies if:
+- it specifies a scope other than `thread_scope_system`, **or**
+
+the scope is `thread_scope_system` and:
```
Comment (daniellustig) [inline docs/extended_api/memory_model.md:52]: Why separate this part out?  Isn't this whole thing logically equivalent to just listing all 5 bullets in a single list?

**Comment by daniellustig** [comment]
Looks good to me, aside from the one minor comment I just posted

File: docs/extended_api/memory_model.md (hunk 6 above)
Comment (gonzalobg) [reply docs/extended_api/memory_model.md:52]: Yes, you are right.  The numbered list below says that this "boolean" expression is true:  1 || (2 && (3 || 4 || 5 || 6)). 
I got feedback that this was hard to follow when it is all together, so I decided to split it when the parenthesis open.

In [rendered](https://github.com/gonzalobg/libcudacxx/blob/docs/memory_model/docs/extended_api/memory_model.md#atomicity) markdown it looked ok to me, but if this can be improved I'd like to improve it.

An atomic operation is atomic at the scope it specifies if:
1. it specifies a scope other than `thread_scope_system`, or
2. the scope is `thread_scope_system` and
3. it affects an object in [unified memory] and [`concurrentManagedAccess`] is `1`, or
4. it affects an object in CPU memory and [`hostNativeAtomicSupported`] is `1`, or
5. it is a load or store that affects a naturally-aligned object of sizes `1`, `2`, `4`, or `8` bytes on [mapped memory], or
6. it affects an object in GPU memory and only GPU threads access it.

**Comment by gonzalobg** [comment]
@wmaxey @daniellustig i think we can just merge this. 
Having this would simplify the docs of atomic_ref as well.

**Comment by daniellustig** [comment]
> @wmaxey @daniellustig i think we can just merge this. Having this would simplify the docs of atomic_ref as well.

👍

**Review by wmaxey — APPROVED** [review]
